lib/Sema/CSDiagnostics.cpp

```diff
@@ -672,10 +672,33 @@ bool GenericArgumentsMismatchFailure::diagnoseAsError() {
         diagnostic = getDiagnosticFor(CTP_AssignSource);
         fromType = getType(assignExpr->getSrc());
         toType = getType(assignExpr->getDest());
+      } else {
+        // Handle cases where we have an apply arg to param
+        if (!path.empty() &&
+            path.front().getKind() == ConstraintLocator::ApplyArgument) {
```

Comment:
You can use `getFunctionArgApplyInfo` instead to get the information about an argument to parameter conversion (which will also handle things like key path subscript arguments).

I think it would also be worth moving this check until after the `switch` so that it's a fallback that can be used more generally if we couldn't find a more specific diagnostic, which I believe would also handle things like an `[Int]` to `[String]?` argument mismatch.

Now that I think about it, it would also be nice to handle more general assignment failures here such as from `[Int]` to `[String]?`, but that can be done separately :)

Reply:
Nice! Thank you @hamishknight 
Didn't know that this `getFunctionArgApplyInfo`, I'll make the adjustments :)

Do you mean for `AssignExpr` in specific? I thought those like `[Int]` to `[String]?` were already handled ...

Reply:
@LucianoPAlmeida Thanks! Could you add a test case for the `[Int]` to `[String]?` argument mismatch example just to make sure it's now covered?

> Do you mean for `AssignExpr` in specific? I thought those like `[Int]` to `[String]?` were already handled ...

Yeah I think we should be able to similarly fallback to an assignment diagnostic if we have an `AssignExpr` anchor, which would supersede the logic we currently have for `case ConstraintLocator::GenericArgument`.

Cases like this currently give a bad diagnostic:

```swift
var x: [Int]?
func foo(_ y: [String]) {
  x = y // error: type of expression is ambiguous without more context
}
```

Reply:
Ahh seems like is the other way around is handled like 
```
var a: [Int]?
var b: [String]

a = b  // error: type of expression is ambiguous without more context
b = a // Already handled
```
I'll adjust here real quick :)

Reply:
@LucianoPAlmeida Awesome! I still think we should add a test case for the `[Int]` to `[String]?` argument mismatch though, what do you think?

Reply:
Sure!

Reply:
Added!